### PR TITLE
TLS support update

### PR DESCRIPTION
**Pull request by 5abeel**
This PR includes:
* Inclusion of CredentialsManager for compilation
* gRPC clients updated to support TLS mode
* Script that assists in setting up certificates in default location
* Documentation
* Update of TIP stratum-dev to include the TLS mode support

### CODE REVIEW

**Review by vsureshkumarp — APPROVED** [review]
Changes looks good to me.

File: docs/ipdk-security.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# IPDK Security
+
+## Overview
+
+This document captures all the details related to gRPC ports and secure/insecure modes and certificate management.
+
+## gRPC Ports
+
+IPDK uses a secure-by-default model. The gRPC server will secure ports by default. A user may choose to open insecure ports if needed, but this will be a conscience decision taken by the user at run time.
```
Comment (ffoulkes) [inline docs/ipdk-security.md:9]: **conscious** decision ("on purpose"). **consciense** means something else entirely.

File: docs/ipdk-security.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+> /usr/share/stratum/certs/
+
+If alternate location is desired, the location will need to be specified during runtime with the following flags
+
+> -ca_cert_file=[CA certificate file]
+> -server_cert_file=[Server certificate file]
+> -server_key_file=[Server private key file]
+
+For e.g.:
```
Comment (ffoulkes) [inline docs/ipdk-security.md:67]: Just say "For example:".

_e.g._ is an abbreviation for the Latin expression _exempli gratia_. It means "for example". _For e.g._ translates literally as "for for example," which is nonsense.

_e.g._ is normally inserted into a longer sentence as part of a supporting phrase. As a lead-in, you want to use the English phrase by itself.

File: docs/ipdk-security.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+### Opening Ports in Insecure Mode
+
+Ports can be opened in insecure mode by user if needed. This is controlled by a flag that needs to be enabled during runtime. Change the grpc_open_insecure_ports value to true to open insecure ports.
+
+To launch InfraP4D with insecure ports 9339 and 9559 open:
+
+```bash
+
```
Comment (ffoulkes) [inline docs/ipdk-security.md:83]: Omit lines 83 and 85. You don't need additional white space around the sample code.

File: docs/ipdk-security.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+> -ca_cert_file=[CA certificate file]
+> -server_cert_file=[Server certificate file]
+> -server_key_file=[Server private key file]
+
+For e.g.:
+
+```bash
+# Files present in /tmp/certs/ directory
+
```
Comment (ffoulkes) [inline docs/ipdk-security.md:71]: Omit lines 71 and 73. You don't need additional vertical white space around the sample code.

File: docs/ipdk-security.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+To launch InfraP4D with insecure ports 9339 and 9559 open:
+
+```bash
+
+$IPDK_RECIPE/install/sbin/infrap4d  -grpc_open_insecure_mode=true
+
+```
+
+
```
Comment (ffoulkes) [inline docs/ipdk-security.md:88]: Just one blank line between sections.

File: docs/ipdk-security.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+## gRPC Clients
+
+Under default conditions, the gRPC clients will require the TLS certificates to establish communication with p4ovs/infrap4d server. The clients will need to use the same ca.crt file and the client.key and client.crt files signed by the ca.crt (can copy the generated files from the server if client is not on the same system as server).
+
+P4RT Client
```
Comment (ffoulkes) [inline docs/ipdk-security.md:93]: Did you mean to make "P4RT Client" a heading (perhaps `### P4RT Client`)?

As written here, the two lines will be run together as "P4RT Client The p4rt-ctl (P4RT client) will default to ...".

File: ovs-p4rt/ovs_p4rt.cc [hunk 7]
```diff
@@ -227,7 +227,7 @@ absl::Status ConfigFdbTunnelTableEntry(ovs_p4rt_cpp::OvsP4rtSession* session,
 void ConfigFdbTableEntry(struct mac_learning_info learn_info, bool insert_entry) {
     // Start a new client session.
     auto status_or_session = ovs_p4rt_cpp::OvsP4rtSession::Create(
-      absl::GetFlag(FLAGS_grpc_addr), ::grpc::InsecureChannelCredentials(),
+      absl::GetFlag(FLAGS_grpc_addr),
```
Comment (ffoulkes) [inline ovs-p4rt/ovs_p4rt.cc:230]: Google coding conventions would place these on the same line.

File: ovs-p4rt/ovs_p4rt_session.cc [hunk 8]
```diff
@@ -84,12 +84,29 @@ absl::StatusOr<std::unique_ptr<OvsP4rtSession>> OvsP4rtSession::Create(
...
 
+std::shared_ptr<::grpc::ChannelCredentials> OvsP4rtSession::getChannelCredentials() {
+  std::string cert, key, ca;
+
+  // TODO(5abeel): Hardcoding default certificate file locations for now.
+  readFile("/usr/share/stratum/certs/ca.crt", ca);
+  readFile("/usr/share/stratum/certs/client.key", key);
+  readFile("/usr/share/stratum/certs/client.crt", cert);
+  
+  grpc::SslCredentialsOptions opts = { ca ,key ,cert };
```
Comment (ffoulkes) [inline ovs-p4rt/ovs_p4rt_session.cc:100]: The space should go _after_ the comma, not _before_ it, just as it does in normal English writing. This spacing is bizarre and jarring. It detracts from the readability of the code.

File: ovs-p4rt/ovs_p4rt_session.h [hunk 9]
```diff
@@ -28,6 +30,18 @@ inline absl::uint128 TimeBasedElectionId() {
   return absl::MakeUint128(absl::ToUnixSeconds(absl::Now()), 0);
 }
 
+static void readFile(const std::string& filename, std::string& data)
```
Comment (ffoulkes) [inline ovs-p4rt/ovs_p4rt_session.h:33]: The static function definition (and the #includes for \<fstream\> and \<sstream\>) are not part of the interface and do _not_ belong in this header file.

You're also not following the coding conventions for this module.

**Pull request merged**